// csv-to-mysql/index.js
var fs = require('fs');
var path = require('path');
var mysql = require('mysql');
var table_name = 'mapdb.input_ARIS'
var date_patt = new RegExp(/^(0?[1-9]|[12][0-9]|3[01])[\/\-](0?[1-9]|1[012])[\/\-]\d{4}$/) ;
var connection; 
function connect(host,db,user,pwd,csvPath)
{
    var host = host;
    var db = db;
    var user = user;
    var pwd = pwd;
    if(!db&&!user&&!pwd)
    {
        console.log('please enter all parameters');
        return;
    }
    else
    {
           connection = mysql.createConnection({
                host     : host,
                user     : user,
                password : pwd
            });
            //.connection.connect();
            
            connection.query(' CREATE DATABASE IF NOT EXISTS '+db+";",function(err){
                 if(err)
                 {
                     console.log(err);
                   
                     return;
                 }
                else
                {
                    connection.query('use '+db+';');
                    readfile(csvPath);
                    
                }
            });
            
   
    }
}
function readfile(csvPath)
{
   var csvPath = csvPath;
   var ext=path.extname(csvPath);
   //table_name=path.basename(csvPath);
   //table_name=table_name.replace(/.csv/,'')
   console.log(table_name);
   
   if(ext!=".csv")
   {
        console.log('Please Select a csv file!');     
        return;
   }
    else
    {
        //insert_table();
        IfExsists(csvPath);
           
    }
}

/*
function CreateTable()
{
    var str="Create Table "+table_name+"(";
    for(i=0;i<count;i++)
    {
      str+=values[i]+" "+findtype(values[count+i]);
        if(i!=count-1)
            str+=","
    }
    str+=");"
    console.log(str);
    connection.query(str);
    insert_table();
}
*/


function IfExsists(csvPath)
/*
{
    str = "select * from "+table_name+";";
    connection.query(str,function(err){
       if(!err)
        {
            console.log("Table already exsists!!!!");
            connection.end();
            return;

            if(choice=='y'||choice=='Y')
            console.log(err);
            
            {
              connection.query('DROP TABLE '+table_name+';');
               fs.readFile(csvPath,function(err,data){
               count = FindColNo(data);
               data = data.toString(); 
               values = data.split(/[,\n]/);
               CreateTable(values);
            });      
            }
            else
            {
                console.log('Rename csv file to create a table with another name!');   
            }
        }
        else
        */

        {
           fs.readFile(csvPath,function(err,data){
           //count = FindColNo(data);
           data = data.toString(); 
           values = data.split(/[,\n]/);
           insert_table()
           //CreateTable(values);
            });     
        }
    //});
//}


function parse_date(datestring)
{
    var parsedString="";
    var arr = [];
    arr = datestring.split("/");
    for(var i=arr.length-1;i>=0;i--)
    {
        parsedString +=arr[i];
        if(i!=0)
            parsedString+='-';
    }
    return parsedString;
}
function insert_table()
{
   j=1;
   while(1)
    {
        var str="Insert into "+table_name+" values (";
        var count = 5;
        for(var i=j*count;i<(j+1)*count;i++)
        {
            if(!values[i])
            {
                console.log("Please don't enter null values!! Check your csv file!!");
                connection.query('DROP TABLE '+table_name+";");
                return;
            }
            
            if(isNaN(values[i]))
            {
                if(date_patt.test(values[i]))
                {
                  var ins_date = parse_date(values[i])
                  str+="'"+ins_date+"'";
//                console.log(values[i]);
//                str+=values[i];
                }
                else
                {
                    str+="'"+values[i]+"'"; 
                }
            }
            else
            {
                str+=values[i]; 
            }
            if(i!=(j+1)*count-1)
                str+=","
        }
        str+=");"
        //console.log(str);
        connection.query(str);
        str="";
        j++;
            if(j>values.length/count-1)
                break;
        
    }
        console.log('Data inserted succesfully!!');
        connection.end();
    
}

/*
function findtype(x)
{
    if(!x)
    {
        console.log("Null values cannot be added into the table!!");
        connection.query('DELETE FROM '+table_name+";");
        return;
    }
    if(!isNaN(x))
    {
        var n = Number(x);
        if(n % 1 === 0)
            return "int";   
        else
            return "float";
    }
   else
   {  
     if(date_patt.test(x))
       {
            return("date");   
       }
           
       return("varchar("+x.length*3+")");
   }
}
*/

/*
function FindColNo(data)
{
    data = data.toString();
    var values= [];
    values = data.split(/[,]/);
    var count = 0;
    for(i=0;i<values.length;i++)
    {
        count++;   
        if(values[i].indexOf('\n')!=-1)
            return count;
    }
}
*/

module.exports = connect;
